### PR TITLE
A visitor for exceeding canvas (and margin) by a complex command

### PR DESCRIPTION
**Pull request by jowczar**

### CODE REVIEW

File: src/main/java/edu/kis/powp/jobs2d/command/CommandBoundariesCheckVisitor.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
+package edu.kis.powp.jobs2d.command;
+
+import javax.swing.*;
+import java.awt.*;
+import java.util.Iterator;
+
+import static java.lang.Math.abs;
+
+public class CommandBoundariesCheckVisitor implements Visitor{
+	private final JPanel canvas;
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/CommandBoundariesCheckVisitor.java]: Add CanvasChecker abstraction and JPanel adapter.

**Review by kowallus — CHANGES_REQUESTED** [review]

File: src/main/java/edu/kis/powp/jobs2d/command/CommandBoundariesCheckVisitor.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+			DriverCommand driverCommand = iterator.next();
+
+			driverCommand.accept(this);
+		}
+	}
+
+	@Override
+	public void visitOperateToCommand(OperateToCommand operateToCommand) {
+		if(abs(operateToCommand.getPosX()) > canvas.getWidth()/2 || abs(operateToCommand.getPosY()) > canvas.getHeight()/2){
+			isExceedingBoundaries = true;
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/CommandBoundariesCheckVisitor.java]: Such implementation of the visitor is not reusable in run-time. Once flag is set it cannot be unset and following accept calls would always return true. 
To fix this you should set the flag properly (to true or false!) in each visit method (also in visitICompoundCommand)

File: src/main/java/edu/kis/powp/jobs2d/command/CommandBoundariesCheckVisitor.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	}
+
+	@Override
+	public void visitICompoundCommand(ICompoundCommand iCompoundCommand) {
+		Iterator<DriverCommand> iterator = iCompoundCommand.iterator();
+
+		while(iterator.hasNext()) {
+			DriverCommand driverCommand = iterator.next();
+
+			driverCommand.accept(this);
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/CommandBoundariesCheckVisitor.java:20]: notice that you can short-cut (break) the visitation if isExceedingBoundaries is set to true.

File: src/main/java/edu/kis/powp/jobs2d/command/ICanvas.java [hunk 4]
```diff
@@ -0,0 +1,6 @@
+package edu.kis.powp.jobs2d.command;
+
+public interface ICanvas {
+	 int getWidth();
+	 int getHeight();
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/ICanvas.java]: Please, do not assume that each canvas wound be rectangular (you can change name to IRectangularCanvas)

File: src/main/java/edu/kis/powp/jobs2d/command/CommandBoundariesCheckVisitor.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+	@Override
+	public void visitOperateToCommand(OperateToCommand operateToCommand) {
+		if(abs(operateToCommand.getPosX()) > canvas.getWidth()/2 || abs(operateToCommand.getPosY()) > canvas.getHeight()/2){
+			isExceedingBoundaries = true;
+		}
+	}
+
+	@Override
+	public void visitSetPositionCommand(SetPositionCommand setPositionCommand) {
+		if(abs(setPositionCommand.getPosX()) > canvas.getWidth()/2 || abs(setPositionCommand.getPosY()) > canvas.getHeight()/2){
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/CommandBoundariesCheckVisitor.java]: Code duplication. You should apply SRP: checking this condition can be delegated to some strategy.

**Review by kowallus — CHANGES_REQUESTED** [review]
This solution can be improved in a few aspects.

File: src/main/java/edu/kis/powp/jobs2d/command/CommandBoundariesCheckVisitor.java [hunk 6]
```diff
@@ -4,11 +4,11 @@
 
 import static java.lang.Math.abs;
 
-public class CommandBoundariesCheckVisitor implements Visitor{
-	private final ICanvas canvas;
+public class CommandBoundariesCheckVisitor implements Visitor, ICommandBoundariesCheckStrategy{
+	private final IRectangularCanvas canvas;
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/CommandBoundariesCheckVisitor.java]: CommandBoundariesCheckVisitor doesn't need to assume work with rectangular canvas. It is an unnecessary limitation.

File: src/main/java/edu/kis/powp/jobs2d/command/CommandBoundariesCheckVisitor.java [hunk 7]
```diff
@@ -4,11 +4,11 @@
 
 import static java.lang.Math.abs;
 
-public class CommandBoundariesCheckVisitor implements Visitor{
-	private final ICanvas canvas;
+public class CommandBoundariesCheckVisitor implements Visitor, ICommandBoundariesCheckStrategy{
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/CommandBoundariesCheckVisitor.java]: Implementing ICommandBoundariesCheckStrategy is closer to the Template Method pattern approach. In Strategy pattern ICommandBoundariesCheckStrategy would be an attribute. Please refactor.

**Review by kowallus — CHANGES_REQUESTED** [review]
It is not how the Strategy pattern is implemented. Refactor.

**Comment by jowczar** [comment]
@kowallus
I made a couple of changes.

– I changed it so now check strategy is an actual staregy pattern. It will hold the logic of calculation of canvas, while Visitor only takes care of executing it on every command.

– I also made Visitor accept any canvas as I don't think it should be limited/determined in the visitor

– I deleted IRectangularCanvas as I didn't see further use for it for now. I think all canvas should have width and height (the outermost points in case of shapes) as it may be helpful in future functionalities so I left those getters in ICanvas. 

If they are in a shape of, for example, a star, they can still use RectangularCanvasCheckStrategy to get "box" calculation or somebody can implement FigureCanvasCheckStrategy and calculate exact field of canvas and whether the provided point is out of it. 

If there is a need for any other weird calculation, I think this is the way it is possible to be extended without changes in our code.

– I wasn't sure about naming and placement of the files. I renamed the ICommandBoundairesCheckStrategy to ICanvasBoundariesCheckStrategy. It can be used anywhere, it doesn't have to be a command-related.

– finally, I moved abs() into calculation function.

LMK if anything needs change

File: src/main/java/edu/kis/powp/jobs2d/command/ICanvasBoundariesCheckStrategy.java [hunk 8]
```diff
@@ -0,0 +1,5 @@
+package edu.kis.powp.jobs2d.command;
+
+public interface ICanvasBoundariesCheckStrategy {
+	boolean checkExceedingBoundaries(ICanvas canvas, int X, int Y);
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/ICanvasBoundariesCheckStrategy.java]: IMO: canvas parameter is not necessary here. ICanvas object can be simply a attribute of ICanvasBoundariesCheckStrategy implementation.

**Review by kowallus — CHANGES_REQUESTED** [review]

**Pull request merged**